src/textual/widgets/text_input.py

```diff
@@ -200,33 +323,63 @@ def on_key(self, event: events.Key) -> None:
+            self._update_suggestion(event)
+        elif key == "ctrl+d":
+            self._update_suggestion(event)
+        elif key == "home" or key == "ctrl+a":
             self.visible_range = (0, available_width)
-        elif key == "end":
-            value_length = len(self.value)
+        elif key == "end" or key == "ctrl+e":
+            num_codepoints = len(self.value)
+            final_visible_codepoints = self._editor.get_range(
+                num_codepoints - available_width + 1,
+                max(num_codepoints, available_width) + 1,
+            )
+            cell_len_final_visible = cell_len(final_visible_codepoints)
+
+            # Additional shift to ensure there's space for double width character
+            additional_shift_required = (
+                max(0, cell_len_final_visible - available_width) + 2
+            )
             if scrollable:
                 self.visible_range = (
-                    value_length - available_width + 1,
-                    max(available_width, value_length) + 1,
+                    num_codepoints - available_width + additional_shift_required,
+                    max(available_width, num_codepoints) + additional_shift_required,
                 )
             else:
                 self.visible_range = (0, available_width)
-        elif event.is_printable:
-            # If we're at the end of the visible range, and the editor backend
-            # will permit us to move the cursor right, then shift the visible
-            # window/range along to the right.
-            if cursor_index == end - 1:
-                self.visible_range = start + 1, end + 1
 
         # We need to clamp the visible range to ensure we don't use negative indexing
         start, end = self.visible_range
         self.visible_range = (max(0, start), end)
 
+    def _update_suggestion(self, event: events.Key) -> None:
+        """Run the autocompleter function, updating the suggestion if necessary"""
+        if self.autocompleter is not None:
+            # TODO: We shouldn't be doing the stuff below here, maybe we need to add
+            #  a method to the editor to query an edit operation?
+            event.prevent_default()
+            super().on_key(event)
+            if self.value:
+                full_suggestion = self.autocompleter(self.value)
+                if full_suggestion:
+                    suffix = full_suggestion[len(self.value) :]
+                    self._suggestion_suffix = suffix
+                else:
+                    self._suggestion_suffix = None
+            else:
+                self._suggestion_suffix = None
+
+    def _toggle_cursor_visible(self):
+        """Manages the blinking of the cursor - ensuring blinking only starts when the
+        user hasn't pressed a key in some time"""
+        if time.monotonic() - self._last_keypress_time > self.cursor_blink_period:
+            self._cursor_blink_visible = not self._cursor_blink_visible
```

Comment:
Is this, or could it be, a Reactive value? In that case you wouldn't need the explicit refresh.

Reply:
The refreshes are always done on idle, so if you set reactive values and call `refresh()` explicitly, Textual will only repaint the screen once.